From 939e4bf807cf3d5fd6ad471f8b4d17726442e3a0 Mon Sep 17 00:00:00 2001
From: Kihwal Lee <kihwal@apache.org>
Date: Mon, 15 Aug 2016 13:13:49 -0500
Subject: [PATCH 1799/2848] HDFS-9696. Garbage snapshot records linger
 forever. Contributed by Kihwal Lee

(cherry picked from commit 83e57e083f2cf6c0de8a46966c5492faeabd8f2a)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestSaveNamespace.java

(cherry picked from commit 4766a3153dd517ac832d4761c884ed88a83a6c09)
(cherry picked from commit 579709fc222d3b1c7d80c3adc037695d615c62d7)

Change-Id: I13c358a6795b14ec7ff19c8aa397480bce5bd9d0
---
 .../server/namenode/FSImageFormatProtobuf.java     |    6 ++-
 .../hdfs/server/namenode/TestSaveNamespace.java    |   41 ++++++++++++++++++++
 2 files changed, 46 insertions(+), 1 deletion(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java
index d4630cc..e546cbf 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java
@@ -455,7 +455,11 @@ private void saveSnapshots(FileSummary.Builder summary) throws IOException {
           this, summary, context, context.getSourceNamesystem());
 
       snapshotSaver.serializeSnapshotSection(sectionOutputStream);
-      snapshotSaver.serializeSnapshotDiffSection(sectionOutputStream);
+      // Skip snapshot-related sections when there is no snapshot.
+      if (context.getSourceNamesystem().getSnapshotManager()
+          .getNumSnapshots() > 0) {
+        snapshotSaver.serializeSnapshotDiffSection(sectionOutputStream);
+      }
       snapshotSaver.serializeINodeReferenceSection(sectionOutputStream);
     }
 
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestSaveNamespace.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestSaveNamespace.java
index 26bddc3..9762e8d 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestSaveNamespace.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestSaveNamespace.java
@@ -624,6 +624,47 @@ public void testSaveNamespaceWithDanglingLease() throws Exception {
     }
   }
 
+  @Test
+  public void testSkipSnapshotSection() throws Exception {
+    MiniDFSCluster cluster = new MiniDFSCluster.Builder(new Configuration())
+        .numDataNodes(1).build();
+    cluster.waitActive();
+    DistributedFileSystem fs = cluster.getFileSystem();
+    OutputStream out = null;
+    try {
+      String path = "/skipSnapshot";
+      out = fs.create(new Path(path));
+      out.close();
+
+      // add a bogus filediff
+      FSDirectory dir = cluster.getNamesystem().getFSDirectory();
+      INodeFile file = dir.getINode(path).asFile();
+      file.addSnapshotFeature(null).getDiffs()
+          .saveSelf2Snapshot(-1, file, null);
+
+      // make sure it has a diff
+      assertTrue("Snapshot fileDiff is missing.",
+          file.getFileWithSnapshotFeature().getDiffs() != null);
+
+      // saveNamespace
+      fs.setSafeMode(SafeModeAction.SAFEMODE_ENTER);
+      cluster.getNameNodeRpc().saveNamespace();
+      fs.setSafeMode(SafeModeAction.SAFEMODE_LEAVE);
+
+      // restart namenode
+      cluster.restartNameNode(true);
+      dir = cluster.getNamesystem().getFSDirectory();
+      file = dir.getINode(path).asFile();
+
+      // there should be no snapshot feature for the inode, when there is
+      // no snapshot.
+      assertTrue("There should be no snapshot feature for this INode.",
+          file.getFileWithSnapshotFeature() == null);
+    } finally {
+      cluster.shutdown();
+    }
+  }
+
   private void doAnEdit(FSNamesystem fsn, int id) throws IOException {
     // Make an edit
     fsn.mkdirs(
-- 
1.7.9.5

